pepdbagent/base_connection.py

```diff
@@ -0,0 +1,130 @@
+from urllib.parse import urlparse
+import logging
+import psycopg2
+from .const import *
+from .exceptions import SchemaError
+
+_LOGGER = logging.getLogger("pepdbagent")
+
+
+class BaseConnection:
+    """
+    A class with base methods, that are used in several classes. e.g. fetch_one or fetch_all
+    """
+
+    def __init__(
+        self,
+        host="localhost",
+        port=5432,
+        database="pep-db",
+        user=None,
+        password=None,
+        dsn=None,
+    ):
+        """
+        Initialize connection to the pep_db database. You can use The basic connection parameters
+        or libpq connection string.
+        :param host: database server address e.g., localhost or an IP address.
+        :param port: the port number that defaults to 5432 if it is not provided.
+        :param database: the name of the database that you want to connect.
+        :param user: the username used to authenticate.
+        :param password: password used to authenticate.
+        :param dsn: libpq connection string using the dsn parameter
+        (e.g. "localhost://username:password@pdp_db:5432")
+        """
+
+        _LOGGER.info(f"Initializing connection to {database}...")
+
+        if dsn is not None:
+            self.pg_connection = psycopg2.connect(dsn)
+            self.db_name = urlparse(dsn).path[1:]
+        else:
+            self.pg_connection = psycopg2.connect(
+                host=host,
+                port=port,
+                database=database,
+                user=user,
+                password=password,
+            )
+            self.db_name = database
+
+        # Ensure data is added to the database immediately after write commands
+        self.pg_connection.autocommit = True
+
+        self._check_conn_db()
+        _LOGGER.info(f"Connected successfully!")
+
+    def commit_to_database(self) -> None:
+        """
+        Commit to database
+        """
+        self.pg_connection.commit()
+
+    def close_connection(self) -> None:
+        """
+        Close connection with database
+        """
+        self.pg_connection.close()
+
+    def __exit__(self):
```

Comment:
Do we need a `__del__` function that will close the connection when the object gets garbage collected?

Reply:
isn't exit doing the same thing? if not  I will add del

Reply:
Im pretty sure, `__exit__` is called in context managers.

```python
with PEPDatabaseAgent as db:
    db.proj.get(...)
```

So, a `__del__` method is nice for cases outside that.

Reply:
ok, you are right, I didn't know about it. Thanks for this comment. I will add __del__ to our classes